src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -773,43 +776,50 @@ private async Task TryNotifyClientErrorAsync(IClientProxy client, string error,
                             throw new InvalidOperationException($"The first set of update operations must always be of type {nameof(RootComponentOperationType.Add)}");
                         }
                     }
-
-                    pendingTasks = new Task[operations.Length];
                 }
 
-                for (var i = 0; i < operations.Length; i++)
+                var quiescenceTask = HandleInboundActivityAsync(async () =>
                 {
-                    var operation = operations[i];
-                    switch (operation.Type)
+                    var pendingTasks = ArrayPool<Task>.Shared.Rent(operations.Length);
+
+                    try
                     {
-                        case RootComponentOperationType.Add:
-                            var task = webRootComponentManager.AddRootComponentAsync(
-                                operation.SsrComponentId,
-                                operation.Descriptor.ComponentType,
-                                operation.Marker.Value.Key,
-                                operation.Descriptor.Parameters);
-                            if (pendingTasks != null)
+                        for (var i = 0; i < operations.Length; i++)
+                        {
+                            var operation = operations[i];
+                            switch (operation.Type)
                             {
-                                pendingTasks[i] = task;
+                                case RootComponentOperationType.Add:
+                                    pendingTasks[i] = webRootComponentManager.AddRootComponentAsync(
+                                        operation.SsrComponentId,
+                                        operation.Descriptor.ComponentType,
+                                        operation.Marker.Value.Key,
+                                        operation.Descriptor.Parameters);
+                                    break;
+                                case RootComponentOperationType.Update:
+                                    pendingTasks[i] = webRootComponentManager.UpdateRootComponentAsync(
```

Comment:
We previously didn't wait for "update" operations to complete before clearing persistent state in the first root component update batch. But it's likely preferable to wait for them to complete before returning back up the inbound activity handler pipeline, hence this change. If others think this change is problematic, we could take another approach (maybe collecting the "add" operations separately and awaiting those).

Reply:
@javiercn has been the one with the most specific opinions about the semantics for persistent state clearing, so hopefully we can get his view and explanation.